### ISSUE TITLE
Add Support for ESLint v9

### ISSUE THREAD

**Issue by Kenneth-Sills** [Kenneth-Sills/eslint-config-airbnb-typescript#15]
Migrated from upstream issue [here](https://github.com/iamturns/eslint-config-airbnb-typescript/issues/331).

----

See the [official migration documentation](https://eslint.org/docs/latest/use/migrate-to-9.0.0).

We probably don't have to do anything on our end (except #14), so we're mostly waiting on [Airbnb upstream](https://github.com/airbnb/javascript/issues/2961) to add support.

**Comment by Kenneth-Sills** [comment]
Since it's been a while since this ticket was made, just wanted to leave an update: we're still waiting on upstream dependencies right now. But while progress has been slow, there [has been progress!](https://github.com/import-js/eslint-plugin-import/issues/2948)

With that in mind, I'll try to find time in the upcoming two weeks to get ready for v9 so once the blockers are resolved, we can immediately make our own release. :smile:

**Comment by JFGHT** [comment]
@Kenneth-Sills Looks like they're done upstream!

**Comment by claabs** [comment]
I've been doing some experimenting with v9 migration. I found that some extra tools like the [Config Inspector](https://eslint.org/blog/2024/04/eslint-config-inspector/) do not support `require`. 

Something to consider when supporting v9 is whether we want to drop legacy config (.eslintrc) support or not. If we want to support legacy config **and** want to support v9 Config Inspector, we'll need to add a build process that compiles an ESM and CJS version of the package.

Summary of the options I brainstormed:

### Pure ESM

✅ No build process
✅ Config inspector support
🚫 No legacy config support
🚫 No `eslint.config.cjs` support

### Build system (ESM + CJS)

✅ Legacy config support
✅ `eslint.config.cjs` support
✅ Config inspector support
🚫 Cannot be directly installed from repo URL

### Pure CJS

✅ No build process
✅ Legacy config support
✅ `eslint.config.cjs` support
🚫 No config inspector support

**Comment by Kenneth-Sills** [comment]
Thanks for bringing this up! While I would have gone with a dual module approach when I started the fork, I now agree with Anthony Fu that [it's time to move on](https://antfu.me/posts/move-on-to-esm-only)[^1].

As of [Node 20.17](https://nodejs.org/en/blog/release/v20.17.0), users are able to [`require()` an ESM module](https://nodejs.org/api/modules.html#loading-ecmascript-modules-using-require). As of [20.19](https://nodejs.org/en/blog/release/v20.19.0), this is enabled by default. And ESLint 8.57 [allows you to use an `eslint.config.mjs` even in a `type: commonjs` project](https://eslint.org/blog/2024/02/eslint-v8.57.0-released/#support-for-eslint.config.mjs-and-eslint.config.cjs). So switching to ESM-only will get us nice things and better defaults, while those who want CJS have a simple[^2] path to get it.

[^1]: He's actually the maintainer of the dual-export example you gave, which has [since moved to ESM only](https://github.com/antfu/eslint-config/releases/tag/v4.0.0). 😄 
[^2]: The supported version of Typescript ESLint [requires Node 18](https://github.com/typescript-eslint/typescript-eslint/blob/main/package.json#L50) already, so an upgrade to 20 shouldn't be too tough, with the only major break being [deprecation of calling `url.parse()` with invalid ports](https://github.com/nodejs/node/pull/45526).

**Comment by stevensacks** [comment]
Many other ESLint plugins have released breaking change major version updates that are no longer compatible with ESLint v8. 

I don’t think it’s worth the extra time to make this backwards compatible. People can use the current version of this library with ESLint v8 (like we are now) until they can update to v9.

**Comment by heath-freenome** [comment]
@Kenneth-Sills Is this still stalled due to this [issue](https://github.com/airbnb/javascript/issues/2961)?

**Comment by Kenneth-Sills** [comment]
@heath-freenome Yes and no! I am well past due owing @claabs two reviews, which will make us prepared for ESLint v9 via #14, so we're not entirely stalled. But we can't *actually* support v9 until upstream does, since this plugin obviously contains their incompatible rules.

@stevensacks I'll see how much work is involved in supporting version 8! If it's not much, I don't really mind having some BC, but I'm not ideological about it. 😄

**Comment by Kenneth-Sills** [comment]
>     * There is an urgent need for v9 compatibility which is the reason you forked it in the first place.
>     * There is no reason to install this fork for any reason other than v9 compatibility.

Not quite, this fork was made because the original was abandoned and does not support either the Stylistic project split or v8 of the Typescript plugin. ESLint v9 is the *next* step, but upstream determines the timeline right now.

>     * Spending any time whatsoever trying to make this fork backwards compatible with v8 is time better spent on getting a v9 compatible version released as soon as possible so people can start testing it, giving feedback, and working through any issues.

Keeping v8 support is only an issue for Airbnb, since they make rules that need to consider API compatibility. For us, it's just one extra "legacy" file. It *shouldn't* be an issue. But if it's any real inconvenience, I'm with you.

I've started on the reviews above, so we'll see soon!

**Comment by basememara** [comment]
What would it take to break away from Airbnb's repo and clone instead of fork? It seems the Airbnb eslint is abandoned so would it be reasonable for us to rally around and contribute to make this repo independent of Airbnb's?

**Comment by heath-freenome** [comment]
@basememara @Kenneth-Sills That seems like a great idea, actually.

**Comment by heath-freenome** [comment]
> What would it take to break away from Airbnb's repo and clone instead of fork? It seems the Airbnb eslint is abandoned so would it be reasonable for us to rally around and contribute to make this repo independent of Airbnb's?

@basememara @Kenneth-Sills That seems like a great idea, actually. The actual question is whether this library maintainer wants to pull the configs and implementation over from the underlying airbnb repos and break the dependence on them. It's not totally abandoned, its just that the single maintainer is so adamant about keeping eslint 7 & 8 support that they won't release eslint 9 until it is all buttoned up.

**Comment by Kenneth-Sills** [comment]
@basememara It's on my list of considerations. I respect that maintainer's work - he's a huge foundational member of FOSS and the Node ecosystem as a whole, and his dedication to BC is commendable - but I agree it *does* lead to frustrating decisions like not moving on to more appropriate dependencies (enhanced resolve, for instance).

Charlie has done fantastic work getting us some unit tests that cover a lot of the upstream requirements, and I hope to (start to) extend on that today. If we can get a comprehensive test suite for upstream, that would be the first step.

Appreciate y'alls feedback!

**Comment by basememara** [comment]
Another project, [eslint-config-airbnb-extended](https://github.com/NishargShah/eslint-config-airbnb-extended), builds on Airbnb's lint rules and has ESLint v9 support. It seems like he diverged from the original Airbnb config and looks really clean, definitely worth considering for inspiration. The author also wrote a nice blog post about it.

**Comment by heath-freenome** [comment]
@Kenneth-Sills How can I support you in getting to eslint 9 support without being tied to the old, stuck on 8 airbnb-base?

**Comment by Kenneth-Sills** [comment]
Convince father time to pour some extra sand in my hourglass. 😉 

I'm truthfully just limited by my work hours and higher priority "real" work. Since upstream hasn't moved and most of the plugins we use haven't mandated an upgrade for updates, we're lucky enough not to be impacted and justify this taking my time.

Where I've currently left off is making updates to the initial testing suite to better cover all positive & negative cases from the AirBnB suite. That would be followed by a review & merge of the PR so graciously provided by Charlie. Finally, work would follow to replace the upstream dependency.

Apologies if my timeline is causing you any trouble!

If it is, @NishargShah has worked hard on their own replacement via [`eslint-config-airbnb-extended`](https://www.npmjs.com/package/eslint-config-airbnb-extended). I haven't audited this to testify on quality, and it [isn't a 1-to-1 migration yet](https://github.com/NishargShah/eslint-config-airbnb-extended/issues/10), but it may be able to unblock you!

**Comment by NishargShah** [comment]
Hey @Kenneth-Sills 

One-to-one replacement configs are ready, please check the latest version ( V2.0.0 )

NPM: https://www.npmjs.com/package/eslint-config-airbnb-extended
PR: https://github.com/NishargShah/eslint-config-airbnb-extended/pull/11

**Comment by heath-freenome** [comment]
> Hey [@Kenneth-Sills](https://github.com/Kenneth-Sills)
> 
> One-to-one replacement configs are ready, please check the latest version ( V2.0.0 )
> 
> NPM: https://www.npmjs.com/package/eslint-config-airbnb-extended PR: [NishargShah/eslint-config-airbnb-extended#11](https://github.com/NishargShah/eslint-config-airbnb-extended/pull/11)

@Kenneth-Sills Does this mean you can switch your dependencies and get to full eslint 9 support?

**Comment by NishargShah** [comment]
> > Hey [@Kenneth-Sills](https://github.com/Kenneth-Sills)
> > One-to-one replacement configs are ready, please check the latest version ( V2.0.0 )
> > NPM: https://www.npmjs.com/package/eslint-config-airbnb-extended PR: [NishargShah/eslint-config-airbnb-extended#11](https://github.com/NishargShah/eslint-config-airbnb-extended/pull/11)
> 
> [@Kenneth-Sills](https://github.com/Kenneth-Sills) Does this mean you can switch your dependencies and get to full eslint 9 support?

Thats right @heath-freenome

**Comment by Kenneth-Sills** [comment]
@heath-freenome Rather, that you can *replace* this project with `eslint-config-airbnb-extended`. The two are not intended to work together, AFAICT.

**Comment by NishargShah** [comment]
> [@heath-freenome](https://github.com/heath-freenome) Rather, that you can _replace_ this project with `eslint-config-airbnb-extended`. The two are not intended to work together, AFAICT.

Thats right

**Comment by heath-freenome** [comment]
> > [@heath-freenome](https://github.com/heath-freenome) Rather, that you can _replace_ this project with `eslint-config-airbnb-extended`. The two are not intended to work together, AFAICT.
> 
> Thats right

@NishargShah @Kenneth-Sills So what you are saying is that I can replace both `eslint-config-airbnb` and `@kesills/eslint-config-airbnb-typescript` with `eslint-config-airbnb-extended` and then use eslint 9?

**Comment by NishargShah** [comment]
> > > [@heath-freenome](https://github.com/heath-freenome) Rather, that you can _replace_ this project with `eslint-config-airbnb-extended`. The two are not intended to work together, AFAICT.
> > 
> > 
> > Thats right
> 
> [@NishargShah](https://github.com/NishargShah) [@Kenneth-Sills](https://github.com/Kenneth-Sills) So what you are saying is that I can replace both `eslint-config-airbnb` and `@kesills/eslint-config-airbnb-typescript` with `eslint-config-airbnb-extended` and then use eslint 9?

@heath-freenome Yes you are right 💯

**Comment by heath-freenome** [comment]
> > > > [@heath-freenome](https://github.com/heath-freenome) Rather, that you can _replace_ this project with `eslint-config-airbnb-extended`. The two are not intended to work together, AFAICT.
> > > 
> > > 
> > > Thats right
> > 
> > 
> > [@NishargShah](https://github.com/NishargShah) [@Kenneth-Sills](https://github.com/Kenneth-Sills) So what you are saying is that I can replace both `eslint-config-airbnb` and `@kesills/eslint-config-airbnb-typescript` with `eslint-config-airbnb-extended` and then use eslint 9?
> 
> [@heath-freenome](https://github.com/heath-freenome) Yes you are right 💯

@NishargShah Awesome... So it seems like this repo could theoretically be archived in favor of yours. I'm assuming that the rules are 1:1 for typescript and the old airbnb repo.

@Kenneth-Sills could you update your README.md to reference this other repo as a eslint 9 replacement?

**Comment by Kenneth-Sills** [comment]
> could you update your README.md to reference this other repo as a eslint 9 replacement?

I was holding off on this, since I don't personally know Nisharg and the project was rather new. But with his project gaining traction and them working a bit with the ESLint devs, I'm more comfortable starting the process of deprecating my library here and pointing them to his!

Thanks again for all your work, @NishargShah! Greatly appreciated. 😄

**Comment by NishargShah** [comment]
Thanks a lot @Kenneth-Sills for your support 🤝🏻
